File: src/main/java/io/vproxy/vfx/entity/input/MouseWheelScroll.java
package io.vproxy.vfx.entity.input;

import vjson.JSON;
import vjson.deserializer.rule.IntRule;
import vjson.deserializer.rule.ObjectRule;
import vjson.deserializer.rule.Rule;
import vjson.deserializer.rule.StringRule;
import vjson.util.ObjectBuilder;

public class MouseWheelScroll {
    public final Direction direction;
    public final int value;

    public static final Rule<MouseWheelScroll> rule = ObjectRule.builder(MouseWheelScrollBuilder::new, MouseWheelScrollBuilder::build, builder -> builder
        .put("direction", (o, it) -> o.direction = Direction.valueOf(it), StringRule.get())
        .put("value", (o, it) -> o.value = it, IntRule.get())
    );

    private static class MouseWheelScrollBuilder {
        Direction direction;
        int value;

        MouseWheelScroll build() {
            return new MouseWheelScroll(direction, value);
        }
    }

    public MouseWheelScroll(Direction direction) {
        this.direction = direction;
        this.value = 0;
    }

    public MouseWheelScroll(Direction direction, int value) {
        this.direction = direction;
        this.value = value;
    }

    public JSON.Object toJson() {
        return new ObjectBuilder()
            .put("direction", direction.name())
            .put("value", value)
            .build();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MouseWheelScroll that = (MouseWheelScroll) o;

        if (value != that.value) return false;
        return direction == that.direction;
    }

    @Override
    public int hashCode() {
        int result = direction.hashCode();
        result = 31 * result + value;
        return result;
    }

    @Override
    public String toString() {
        if (value == 0) {
            return "scroll-" + direction.name().toLowerCase();
        } else {
            return "scroll-" + direction.name().toLowerCase() + ":" + value;
        }
    }

    public enum Direction {
        UP,
        DOWN,
    }
}

File: src-ts/src/entity/input/MouseWheelScroll.ts
export enum MouseWheelDirection {
  UP = 'UP',
  DOWN = 'DOWN',
}

export class MouseWheelScroll {
  constructor(public readonly direction: MouseWheelDirection, public readonly value: number = 0) {}

  toJson(): { direction: string; value: number } {
    return { direction: this.direction, value: this.value };
  }

  static fromJson(obj: { direction: string; value: number }): MouseWheelScroll {
    return new MouseWheelScroll(obj.direction as MouseWheelDirection, obj.value);
  }

  equals(o: unknown): boolean {
    return (
      o instanceof MouseWheelScroll &&
      o.direction === this.direction &&
      o.value === this.value
    );
  }

  hashCode(): number {
    let h = hashStr(this.direction);
    h = 31 * h + this.value;
    return h;
  }

  toString(): string {
    const dir = this.direction.toLowerCase();
    if (this.value === 0) return `scroll-${dir}`;
    return `scroll-${dir}:${this.value}`;
  }
}

function hashStr(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) {
    h = (31 * h + s.charCodeAt(i)) | 0;
  }
  return h;
}
